JoinColony/colonyCDapp
# Feat: Refactor manage permissions motion to handle multi-sig

Related block-ingestor PR: https://github.com/JoinColony/block-ingestor/pull/247

## Description

As a user moderating a colony, I want to be able to create and finalize a multi-sig manage roles motion.

It should be possible to create a manage permissions multi-sig motion in the general team. Users with the `root` role (owner) should be able to create, vote and finalize.

It should be possible to create a manage permissions multi-sig motion in any sub team. Users with the `architecture` role (admin) should be able to create, vote and finalize.

(At a later point, in a sub team it should be possible to vote if you have either the `root` or `architecture` role in `general`, but this will be addressed in a separate PR.)

(Note: The completed action screen will not show the correct permissions if you are updating the multi-sig permissions of someone who already has multi-sig permissions. The same issue occurs on master when assigning permissions via a reputation motion so I've opened a new issue for this.)

## Testing

* Step 1: Install the multi-sig extension. Set the fixed threshold to 2 in the extension settings tab.
* Step 2: Assign admin permissions to Amy using `Take actions via Multi-Sig` as the authority, and `Permissions` as the decision method.

<img width="681" alt="Screenshot 2024-07-05 at 15 16 44" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/7c8c410f-a26e-417c-a90e-2154e3c0a5b9">

* Step 3: Assign admin permissions to Fry in the `Andromeda` team using `Take actions via Multi-Sig` as the authority, `Multi-Sig` as the decision method, and `Andromeda` as the `CreatedIn` domain.

<img width="681" alt="Screenshot 2024-07-05 at 15 18 29" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/592111bc-8572-477c-bbba-851a6e102a12">

<img width="1065" alt="Screenshot 2024-07-05 at 15 19 48" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/68fd8d6b-f64a-4c34-b78d-1e5e2678a9a8">

* Step 4: Open an incognito window and connect Amy's wallet. Copy the URL to open the same multi-sig transaction in the incognito window. Then approve the multi-sig motion as Amy.

<img width="1059" alt="Screenshot 2024-07-05 at 15 21 42" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/c80bf27f-edb2-42cc-8d6a-086b2ad5a13b">

* Step 5: Finalize the multi-sig motion. Refresh and confirm Fry has been assigned the appropriate permissions.

<img width="1288" alt="Screenshot 2024-07-05 at 15 22 39" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/17bb6654-0a7e-4d2e-ac69-8198a66cbf20">

* Further testing: Step 6: Switch to Fry's wallet and create a manage permissions action using your new found permission powers. Select `Multi-Sig` as the decision method and `General` as the team. Note both validation notifications. As Fry, we don't have multi-sig permissions in `General` so we see the permissions notification, but also Leela is the only user with `Root` permissions in `General` so there are also not enough users with permissions for the action. When managing permissions in the `root` domain, the `root` permission role is required.

<img width="1290" alt="Screenshot 2024-07-05 at 15 37 32" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/8f93431a-e480-4713-a7d6-204d10ef3dd2">

* Step 7: Switch the team to `Serenity`. When managing permissions in a sub-domain, only the `architecture` role is required. Amy is an admin in `General` so she has the `architecture` role meaning there are sufficient members for the multi-sig motion to be created.

* Step 8: Switch the team to `Andromeda`. Fry actually has permissions here so we can create the multi-sig motion. Leela and Amy both have permissions in General so we should expect to see them as eligible signees here. Select a member, and a permissions type. Set the authority to `Take actions on their own` so we can test this works okay too.

<img width="1293" alt="Screenshot 2024-07-05 at 15 46 24" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/ece4009c-291d-4f61-8598-9ef5ee8e4eff">

<img width="1290" alt="Screenshot 2024-07-05 at 15 46 37" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/dd34aa8b-5347-466e-9459-9afbaa29b9e7">

* Step 9: Switch back to Leela or Amy to approve and finalize.

<img width="1716" alt="Screenshot 2024-07-05 at 15 47 54" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/cce62d67-97ca-41b2-ae58-cfd750f69b62">

Really further testing:
* Check you can assign a user permissions in a sub-domain with `createdIn` set to general. (You will need to assign owner permissions to Amy in `General` to have enough users to vote this through).
* Install the reputation voting extension and check motions are unaffected by any changes here. (Test in general, sub-domain, sub-domain created in general - can also test assigning own permissions or multi-sig permissions).
* Can also test multi-sig motions with both the multi-sig extension and reputation extension installed.
* Can also test reputation motions with the multi-sig extension uninstalled.

## Diffs

**New stuff** ✨

* Manage permissions motion now supports multi-sig decision method
* Manage permissions motion now supports assigning multi-sig permissions via multi-sig decision method

## TODO

- [ ] It should be possible to create / vote as a user with either the `root` role or `architecture` role in a sub-domain. Currently it requires `architecture` and users with only `root` will be unable to create or vote. I will address in this a future PR.

Contributes #2192
